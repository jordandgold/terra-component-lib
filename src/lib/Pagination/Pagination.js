import React, { Component } from "react";
import "./Pagination.scss";
import PropTypes from "prop-types";
import Icon from "../Icon/Icon";

class Pagination extends Component {
  generatePageListItems = () => {
    const totalPages = this.props.pages; //totalPages.length()
    const activePage = this.props.activePage + 1;

    let pages = [];

    if (totalPages === 2) {
      for (let i = 0; i < 2; i++) {
        pages.push(
          <Page
            name={this.props.name}
            key={`list-item-${i + 1}`}
            activePage={activePage}
            page={i + 1}
            handleClick={this.props.handleClick}
          />
        );
      }
    }

    if (totalPages === 3) {
      for (let i = 0; i < 3; i++) {
        pages.push(
          <Page
            name={this.props.name}
            key={`list-item-${i + 1}`}
            activePage={activePage}
            page={i + 1}
            handleClick={this.props.handleClick}
          />
        );
      }
    }

    if (totalPages > 3 && activePage === 1) {
      for (let i = 0; i < 3; i++) {
        pages.push(
          <Page
            name={this.props.name}
            key={`list-item-${i + 1}`}
            activePage={activePage}
            page={i + 1}
            handleClick={this.props.handleClick}
          />
        );
      }
    } else if (totalPages > 3 && activePage === totalPages) {
      for (let i = totalPages - 3; i < totalPages; i++) {
        pages.push(
          <Page
            name={this.props.name}
            key={`list-item-${i + 1}`}
            activePage={activePage}
            page={i + 1}
            handleClick={this.props.handleClick}
          />
        );
      }
    } else if (totalPages > 3 && activePage === totalPages - 1) {
      for (let i = totalPages - 3; i < totalPages; i++) {
        pages.push(
          <Page
            name={this.props.name}
            key={`list-item-${i + 1}`}
            activePage={activePage}
            page={i + 1}
            handleClick={this.props.handleClick}
          />
        );
      }
    } else if (totalPages > 3 && activePage <= totalPages) {
      for (let i = activePage - 2; i < activePage + 1; i++) {
        pages.push(
          <Page
            name={this.props.name}
            key={`list-item-${i + 1}`}
            activePage={activePage}
            page={i + 1}
            handleClick={this.props.handleClick}
          />
        );
      }
    }

    return pages;
  };

  render() {
    return (
      <nav className="ter-pagination" role="navigation">
        <ul className="ter-pagination__list">
          {this.props.activePage + 1 > 1 && this.props.activePage > 0 && (
            <li
              onClick={() =>
                this.props.handleClick(
                  this.props.activePage - 1,
                  this.props.name
                )
              }
              className="ter-pagination__list-item ter-pagination__list-item--previous"
            >
              <Icon type="open-caret-left-dark-8px" size="8px" />
            </li>
          )}
          {this.generatePageListItems()}
          {this.props.activePage + 1 < this.props.pages && (
            <li
              onClick={() => this.props.handleClick(this.props.activePage + 1)}
              className="ter-pagination__list-item ter-pagination__list-item--next"
            >
              <Icon type="open-caret-right-dark-8px" size="8px" />
            </li>
          )}
        </ul>
      </nav>
    );
  }
}

export default Pagination;

Pagination.propTypes = {
  pages: PropTypes.number.isRequired,
  activePage: PropTypes.number.isRequired,
  handleClick: PropTypes.func.isRequired,
  name: PropTypes.string
};

export class Page extends Component {
  handleClick = () => {
    if (this.props.activePage === this.props.page) {
      return;
    } else {
      this.props.handleClick(this.props.page - 1, this.props.name);
    }
  };

  render() {
    const { activePage, page } = this.props;
    return (
      <li
        onClick={this.handleClick}
        className={`ter-pagination__list-item ${
          activePage === page ? "is-active" : ""
        }`}
      >
        {page}
      </li>
    );
  }
}
