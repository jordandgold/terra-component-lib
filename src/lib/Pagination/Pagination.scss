@import "../global-styles/terra.scss";

.ter-pagination {
  &__list {
    list-style: none;
    display: flex;
    text-align: center;
    justify-content: center;
  }
  &__list-item {
    margin: 0;
    padding: spacing("micro") spacing("xxs");
    font-size: font-size("sm");
    &:hover {
      cursor: pointer;
      border-bottom: 2px solid $pagination-border-hover-color;
    }
    &.is-active {
      border-bottom: 2px solid $pagination-border-active-color;
    }
  }
}
