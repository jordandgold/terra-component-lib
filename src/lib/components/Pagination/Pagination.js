import React, { Component } from "react";
import "./Pagination.scss";
import PropTypes from "prop-types";

class Pagination extends Component {
  generatePageListItems = () => {
    return this.props.pages.map((page, index) => {
      return (
        <li key={`list-item-${index}`} className="ter-pagination__list-item">
          <a
            className={`ter-pagination__list-item ${
              page.isActive ? "is-active" : ""
            }`}
            href={`/${page.link}`}
          >
            {page.pageNumber}
          </a>
        </li>
      );
    });
  };

  render() {
    return (
      <nav className="ter-pagination" role="navigation">
        <ul className="ter-pagination__list">
          <li className="ter-pagination__list-item">
            <a href="/" rel="prev">
              <i className="fas fa-chevron-left" />
            </a>
          </li>
          {this.generatePageListItems}
          <li className="ter-pagination__list-item">
            <a href="/" rel="next">
              <i className="fas fa-chevron-right" />
            </a>
          </li>
        </ul>
      </nav>
    );
  }
}

export default Pagination;
