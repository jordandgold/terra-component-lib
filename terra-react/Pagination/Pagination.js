import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import React, { Component } from "react";
import "./Pagination.scss";
import Icon from "../Icon/Icon";

var Pagination =
/*#__PURE__*/
function (_Component) {
  _inherits(Pagination, _Component);

  function Pagination() {
    var _getPrototypeOf2;

    var _this;

    _classCallCheck(this, Pagination);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _possibleConstructorReturn(this, (_getPrototypeOf2 = _getPrototypeOf(Pagination)).call.apply(_getPrototypeOf2, [this].concat(args)));

    _this.generatePageListItems = function () {
      var totalPages = _this.props.pages; //totalPages.length()

      var activePage = _this.props.activePage + 1;
      var pages = [];

      if (totalPages === 2) {
        for (var i = 0; i < 2; i++) {
          pages.push(React.createElement(Page, {
            name: _this.props.name,
            key: "list-item-".concat(i + 1),
            activePage: activePage,
            page: i + 1,
            handleClick: _this.props.handleClick
          }));
        }
      }

      if (totalPages === 3) {
        for (var _i = 0; _i < 3; _i++) {
          pages.push(React.createElement(Page, {
            name: _this.props.name,
            key: "list-item-".concat(_i + 1),
            activePage: activePage,
            page: _i + 1,
            handleClick: _this.props.handleClick
          }));
        }
      }

      if (totalPages > 3 && activePage === 1) {
        for (var _i2 = 0; _i2 < 3; _i2++) {
          pages.push(React.createElement(Page, {
            name: _this.props.name,
            key: "list-item-".concat(_i2 + 1),
            activePage: activePage,
            page: _i2 + 1,
            handleClick: _this.props.handleClick
          }));
        }
      } else if (totalPages > 3 && activePage === totalPages) {
        for (var _i3 = totalPages - 3; _i3 < totalPages; _i3++) {
          pages.push(React.createElement(Page, {
            name: _this.props.name,
            key: "list-item-".concat(_i3 + 1),
            activePage: activePage,
            page: _i3 + 1,
            handleClick: _this.props.handleClick
          }));
        }
      } else if (totalPages > 3 && activePage === totalPages - 1) {
        for (var _i4 = totalPages - 3; _i4 < totalPages; _i4++) {
          pages.push(React.createElement(Page, {
            name: _this.props.name,
            key: "list-item-".concat(_i4 + 1),
            activePage: activePage,
            page: _i4 + 1,
            handleClick: _this.props.handleClick
          }));
        }
      } else if (totalPages > 3 && activePage <= totalPages) {
        for (var _i5 = activePage - 2; _i5 < activePage + 1; _i5++) {
          pages.push(React.createElement(Page, {
            name: _this.props.name,
            key: "list-item-".concat(_i5 + 1),
            activePage: activePage,
            page: _i5 + 1,
            handleClick: _this.props.handleClick
          }));
        }
      }

      return pages;
    };

    return _this;
  }

  _createClass(Pagination, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      return React.createElement("nav", {
        className: "ter-pagination",
        role: "navigation"
      }, React.createElement("ul", {
        className: "ter-pagination__list"
      }, this.props.activePage + 1 > 1 && this.props.activePage > 0 && React.createElement("li", {
        onClick: function onClick() {
          return _this2.props.handleClick(_this2.props.activePage - 1, _this2.props.name);
        },
        className: "ter-pagination__list-item ter-pagination__list-item--previous"
      }, React.createElement(Icon, {
        type: "open-caret-left-dark-8px",
        size: "8px"
      })), this.generatePageListItems(), this.props.activePage + 1 < this.props.pages && React.createElement("li", {
        onClick: function onClick() {
          return _this2.props.handleClick(_this2.props.activePage + 1);
        },
        className: "ter-pagination__list-item ter-pagination__list-item--next"
      }, React.createElement(Icon, {
        type: "open-caret-right-dark-8px",
        size: "8px"
      }))));
    }
  }]);

  return Pagination;
}(Component);

export default Pagination;
export var Page =
/*#__PURE__*/
function (_Component2) {
  _inherits(Page, _Component2);

  function Page() {
    var _getPrototypeOf3;

    var _this3;

    _classCallCheck(this, Page);

    for (var _len2 = arguments.length, args = new Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
      args[_key2] = arguments[_key2];
    }

    _this3 = _possibleConstructorReturn(this, (_getPrototypeOf3 = _getPrototypeOf(Page)).call.apply(_getPrototypeOf3, [this].concat(args)));

    _this3.handleClick = function () {
      if (_this3.props.activePage === _this3.props.page) {
        return;
      } else {
        _this3.props.handleClick(_this3.props.page - 1, _this3.props.name);
      }
    };

    return _this3;
  }

  _createClass(Page, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          activePage = _this$props.activePage,
          page = _this$props.page;
      return React.createElement("li", {
        onClick: this.handleClick,
        className: "ter-pagination__list-item ".concat(activePage === page ? "is-active" : "")
      }, page);
    }
  }]);

  return Page;
}(Component);