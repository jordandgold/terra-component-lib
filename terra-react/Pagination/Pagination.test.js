import React from "react";
import { shallow } from "enzyme";
import Pagination, { Page } from "./Pagination";
describe("Pagination", function () {
  var wrapper;
  var mockPages = 2;
  var mockActivePage = 0;
  var mockHandleClick;
  beforeEach(function () {
    mockHandleClick = jest.fn();
    wrapper = shallow(React.createElement(Pagination, {
      pages: mockPages,
      activePage: mockActivePage,
      handleClick: mockHandleClick,
      name: "name"
    }));
  });
  it("should match the snapshot", function () {
    expect(wrapper).toMatchSnapshot();
  });
  it("should call handleClick on click - forward", function () {
    wrapper.find(".ter-pagination__list-item.ter-pagination__list-item--next").simulate("click");
    expect(mockHandleClick).toHaveBeenCalled();
  });
  it("should call handleClick on click - back", function () {
    mockHandleClick = jest.fn();
    wrapper = shallow(React.createElement(Pagination, {
      name: "name",
      pages: 4,
      activePage: 3,
      handleClick: mockHandleClick
    }));
    wrapper.find(".ter-pagination__list-item").simulate("click");
    expect(mockHandleClick).toHaveBeenCalled();
  });
  describe("generatePageListItems", function () {
    it("should return some JSX if pages === 2", function () {
      var expected = ['<li class="ter-pagination__list-item is-active">1</li>', '<li class="ter-pagination__list-item ">2</li>'];
      var result = wrapper.instance().generatePageListItems().map(function (item) {
        return shallow(item).html();
      });
      expect(result).toEqual(expected);
    });
    it("should return some JSX if pages === 3", function () {
      mockHandleClick = jest.fn();
      wrapper = shallow(React.createElement(Pagination, {
        name: "name",
        pages: 3,
        activePage: mockActivePage,
        handleClick: mockHandleClick
      }));
      var expected = ['<li class="ter-pagination__list-item is-active">1</li>', '<li class="ter-pagination__list-item ">2</li>', '<li class="ter-pagination__list-item ">3</li>'];
      var result = wrapper.instance().generatePageListItems().map(function (item) {
        return shallow(item).html();
      });
      expect(result).toEqual(expected);
    });
    it("should return some JSX if pages > 3 && activePage === 1", function () {
      mockHandleClick = jest.fn();
      wrapper = shallow(React.createElement(Pagination, {
        name: "name",
        pages: 4,
        activePage: mockActivePage,
        handleClick: mockHandleClick
      }));
      var expected = ['<li class="ter-pagination__list-item is-active">1</li>', '<li class="ter-pagination__list-item ">2</li>', '<li class="ter-pagination__list-item ">3</li>'];
      var result = wrapper.instance().generatePageListItems().map(function (item) {
        return shallow(item).html();
      });
      expect(result).toEqual(expected);
    });
    it("should return some JSX if pages > 3 && activePage === total pages", function () {
      mockHandleClick = jest.fn();
      wrapper = shallow(React.createElement(Pagination, {
        pages: 4,
        activePage: 3,
        handleClick: mockHandleClick
      }));
      var expected = ['<li class="ter-pagination__list-item ">2</li>', '<li class="ter-pagination__list-item ">3</li>', '<li class="ter-pagination__list-item is-active">4</li>'];
      var result = wrapper.instance().generatePageListItems().map(function (item) {
        return shallow(item).html();
      });
      expect(result).toEqual(expected);
    });
    it("should return some JSX if pages > 3 && activePage === total pages - 1", function () {
      mockHandleClick = jest.fn();
      wrapper = shallow(React.createElement(Pagination, {
        pages: 4,
        activePage: 2,
        handleClick: mockHandleClick
      }));
      var expected = ['<li class="ter-pagination__list-item ">2</li>', '<li class="ter-pagination__list-item is-active">3</li>', '<li class="ter-pagination__list-item ">4</li>'];
      var result = wrapper.instance().generatePageListItems().map(function (item) {
        return shallow(item).html();
      });
      expect(result).toEqual(expected);
    });
    it("should return some JSX if pages > 3 && activePage <= totalPages", function () {
      mockHandleClick = jest.fn();
      wrapper = shallow(React.createElement(Pagination, {
        pages: 4,
        name: "name",
        activePage: 1,
        handleClick: mockHandleClick
      }));
      var expected = ['<li class="ter-pagination__list-item ">1</li>', '<li class="ter-pagination__list-item is-active">2</li>', '<li class="ter-pagination__list-item ">3</li>'];
      var result = wrapper.instance().generatePageListItems().map(function (item) {
        return shallow(item).html();
      });
      expect(result).toEqual(expected, "name");
    });
  });
  describe("Page", function () {
    var pageWrapper;
    var mockProcessClick;
    beforeEach(function () {
      mockProcessClick = jest.fn();
      pageWrapper = shallow(React.createElement(Page, {
        name: "name",
        activePage: false,
        page: 3,
        handleClick: mockProcessClick
      }));
    });
    it("should match the snapshot", function () {
      expect(pageWrapper).toMatchSnapshot();
    });
    describe("handleClick", function () {
      it("should call handleClick on click", function () {
        var spy = jest.spyOn(pageWrapper.instance(), "handleClick");
        pageWrapper.instance().forceUpdate();
        pageWrapper.find(".ter-pagination__list-item").simulate("click");
        expect(spy).toHaveBeenCalled();
      });
      it("should call processClick if not the active page", function () {
        var expected = 2;
        pageWrapper.instance().handleClick();
        expect(mockProcessClick).toHaveBeenCalledWith(expected, "name");
      });
      it("should return if the active page", function () {
        pageWrapper = shallow(React.createElement(Page, {
          activePage: 1,
          page: 1,
          handleClick: mockProcessClick
        }));
        var spy = jest.spyOn(pageWrapper.instance(), "handleClick");
        pageWrapper.instance().handleClick();
        expect(spy).toHaveReturned();
      });
    });
  });
});