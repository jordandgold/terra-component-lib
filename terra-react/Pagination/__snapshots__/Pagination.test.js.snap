// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination Page should match the snapshot 1`] = `
<li
  className="ter-pagination__list-item "
  onClick={[Function]}
>
  3
</li>
`;

exports[`Pagination should match the snapshot 1`] = `
<nav
  className="ter-pagination"
  role="navigation"
>
  <ul
    className="ter-pagination__list"
  >
    <Page
      activePage={1}
      handleClick={[MockFunction]}
      key="list-item-1"
      name="name"
      page={1}
    />
    <Page
      activePage={1}
      handleClick={[MockFunction]}
      key="list-item-2"
      name="name"
      page={2}
    />
    <li
      className="ter-pagination__list-item ter-pagination__list-item--next"
      onClick={[Function]}
    >
      <Icon
        size="8px"
        type="open-caret-right-dark-8px"
      />
    </li>
  </ul>
</nav>
`;
